records/2024-06-04.manifests-kustomize-cicd.md

```diff
@@ -0,0 +1,140 @@
+
+## Current State
+
+### Staging
+
+#### Individual Application Rollouts
+
+When deploying a new version of the application components (admin/api/document-download etc), the release to staging is kicked off at the application repository level when the docker image is built. The docker image tag is based on the first 7 digits of the github runner sha, and also tags this build as latest. The docker build and push workflow invokes a corresponding workflow in the notification-manifests repository passing the new docker tag to have it deployed to staging.
+
+The manifests rollout workflow receives the new image tag and performs a manual patch of the existing deployment:
+
+```shell
+kubectl set image deployment.apps/api api=$DOCKER_SLUG:$DOCKER_TAG -n=notification-canada-ca
+```
+
+##### Diagram
+
+![Staging Current CICD](./diagrams/2024-06-04.manifests-current-cicd/staging-cicd.png)
+
+#### Global Manifests Rollouts
+
+When a global manifests rollout is performed by the merge to main (staging) github workflow, all of the Kustomize code is deployed.
+
+```shell
+notification-manifests/env/staging kubectl apply -k .
+```
+
+### Production
+
+#### Individual Application Rollouts
+
+This concept does not exist for production. We only rollout all of Kustomize.
+
+#### Global Manifests Rollouts
+
+Every night, the Notify PR Bot reads the latest github commit sha's for each repository, and if they've changed, creates a pull request that changes the image tag values in the Kustomization.yaml for production. Once this is merged, all of the kustomize code is deployed at once.
+
+Since the PR bot does not touch the helmfile side of things, we also do a single global apply on helmfile with "latest" tags.
+
+#### Diagram
+
+![Production Current CICD](./diagrams/2024-06-04.manifests-current-cicd/production-cicd.png)
+
+### Problems
+
+#### Two Different Deployment Technologies
+
+We are currently managing two different Kubernetes deployment methods with Kustomize and Helmfile. It would be beneficial to have everything under one technology.
+
+#### Staging Releases
+
+Because we do not actually update any manifests code when doing an individual application rollout, we lose all context of which version we should be using when we do a global manifest rollout with merge to main (staging). To overcome this, the staging Kustomize code is set to always use the "latest" tag. This means when we apply kustomize the explicitly versioned docker tag is overwritten with "latest". This works because the docker build and push job for each application tags as both latest and the github sha. It is however not ideal since it becomes difficult to track versions in staging, and also creates a delta between how staging and production releases are handled through release pipelines.
```

Comment:
Interesting note to mark it as an issue... as we could say it is part of the mainline/trunk based development paradigm. Switching to versions in staging could align with our previous talks of getting specific releases in staging and using an extra environment such as dev for hosting `latest` mainline. Is this something we want to change and scope within this ADR? Genuinely asking, to know what we are after.